File: src/contests/cg/puzzle/clojure/mirrors.clj
;; 🎮 CodinGame Puzzle - mirrors
;; https://www.codingame.com/training/easy/mirrors

(ns Solution (:gen-class))

(defn -main [& _]
  (read-line)
  (let [r (mapv #(Double/parseDouble %) (.split (.trim (read-line)) "\\s+"))]
    (loop [i (dec (count r)) reflected 0.0]
      (if (>= i 0)
        (let [ri (nth r i)
              denom (- 1 (* ri reflected))
              nr (if (== denom 0) ri (+ ri (/ (* (- 1 ri) (- 1 ri) reflected) denom)))]
          (recur (dec i) nr))
        (println (format "%.4f" reflected)))))
  (flush))

File: src/contests/cg/puzzle/c/mirrors.c
// 🎮 CodinGame Puzzle - mirrors
// https://www.codingame.com/training/easy/mirrors

#include <stdio.h>

int main() {
    int n;
    scanf("%d", &n);
    double r[n];
    for (int i = 0; i < n; i++) scanf("%lf", &r[i]);
    double reflected = 0.0;
    for (int i = n - 1; i >= 0; i--) {
        double ri = r[i];
        double denom = 1 - ri * reflected;
        reflected = ri + (denom == 0 ? 0 : ((1 - ri) * (1 - ri) * reflected) / denom);
    }
    printf("%.4f\n", reflected);
    return 0;
}
